Services/AssessmentQuestion/interfaces/interface.ilAsqResultCalculator.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+/* Copyright (c) 1998-2013 ILIAS open source, Extended GPL, see docs/LICENSE */
+
+/**
+ * Class ilAsqResultCalculator
+ *
+ * @author    Björn Heyser <info@bjoernheyser.de>
+ * @version    $Id$
+ *
+ * @package    Services/AssessmentQuestion
+ */
+interface ilAsqResultCalculator
+{
+	/**
+	 * @param ilAsqQuestion $question
+	 * @return void
+	 */
+	public function setQuestion(ilAsqQuestion $question);
```

Comment:
I am not totally happy with this because I anticipate a temporal coupling problem here.
This interface suggests that `calculate` has to be called by the consumer always **before** `getPoints` and `isCorrect`. IMHO this is error prone. Instead `calculate` could return some `Result` containing points and status. Furthermore something so important like delegating and retrieving results calculations should maybe be worth to be implemented with an [immutable approach](https://en.wikipedia.org/wiki/Immutable_object).
Do we want *Question* and *Solution* to be mutated separately via setters? This could be maybe a bit fragile, because there is no guarantee that the question corresponds to the solution.

Reply:
> This could be maybe a bit fragile, because there is no guarantee that the question corresponds to the solution.

... or the calculated points

I agree with mjansen. I am not sure if I get the scope of the interface completely. In which context is this implemented? Where will instances of it be attached? 

Maybe change to something like, could be handy for consumers:

```
 $myAsqResultCalculator = myAsqFactory->ResultCalculatorX($myQuestion,$mySolution)
        ->withCalculation(new function(){
                //someCalcMagic, maybe this also could be further specified with some calculate instance
         });
```
To me all those different calculators need are different calculations, right? Those could be lazily evaluated if getPoints is called. However, I am really not sure if I get this right...

Reply:
Each question type will come with an own calculator that is able to handle an the qtype corresponding instances of ilAsqQuestion and ilAsqSolution. So calculation does not need to be injected by a function. This additional opportunity to inject consumer depending calculation rules seems to useful. I just leave this out for the moment.

I added an ilAsqQuestionResult as return type for the calculate method. This resolves the temporal coupling.

The Calculator also need to check of the concrete types of ilAsqQuestion and ilAsqSolution of course and it will throw invalid argument exceptions.